recap/src/Imp/Eval.hs

```diff
@@ -11,20 +13,87 @@ import Imp.Lang
 -- Variable mapping; shadowing is allowed
 type VarMap = M.Map String Int
 
--- Possible errors reported during evaluation 
+-- Possible errors reported during evaluation
 data Error = UndefinedVar String | ParsingErr String | DivByZero deriving (Show)
 
 -- Evaluation monad
 type EvalM = StateT VarMap (ExceptT Error IO)
 
+getMap :: String -> EvalM Int
+getMap varName = do
+    varMap <- get
+    case M.lookup varName varMap of
+        Just value -> return value
+        Nothing    -> throwError (UndefinedVar varName)
+
+setMap :: String -> Int -> EvalM ()
+setMap varName value = modify (M.insert varName value)
+
+readInt :: EvalM Int
+readInt = do
+    input <- liftIO getLine
+    case readMaybe input of
+        Just n  -> return n
+        Nothing -> throwError (ParsingErr "expected Int")
+
+readSafe str =
+  case readMaybe str of
+    Just n -> return n
+    Nothing -> throwError (ParsingErr "expected Int")
+
 evalExpr :: Expr -> EvalM Int
-evalExpr = undefined 
+evalExpr expr = do
+  case expr of
+    Var str -> do -- a variable; if it's not defined in the varMap -- throw an exception
+      val <- getMap str
+      return (val)
+    Const val -> do -- a constant
+      return (val)
+    BinOp Plus expr1 expr2 -> do -- a binary operators
+      x <- evalExpr expr1
+      y <- evalExpr expr2
+      return (x + y)
+    BinOp Minus expr1 expr2 -> do
+      x <- evalExpr expr1
+      y <- evalExpr expr2
+      return (x - y)
+    BinOp Mult expr1 expr2 -> do
+      x <- evalExpr expr1
+      y <- evalExpr expr2
+      return (x * y)
+    BinOp Div expr1 expr2 -> do
+      x <- evalExpr expr1
+      y <- evalExpr expr2
+      if y == 0
+        then throwError DivByZero
+        else return (x `div` y)
```

Comment:
You can reduce code duplication by pattern matching on the operator in the body of the function instead.